File: platform/macos/Bridge/NativeShortcuts.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import TidyVNC

public struct NativeShortcutModifiers: OptionSet, Sendable, Hashable {
  public let rawValue: UInt32
  public init(rawValue: UInt32) { self.rawValue = rawValue }
  public static let control = Self(rawValue: 1), shift = Self(rawValue: 2)
  public static let option = Self(rawValue: 4), command = Self(rawValue: 8)
  public static let builtIn: Self = [.control, .option]
}
public enum NativeShortcutAction: UInt32, Sendable { case normal, unarm, shortcut, ignore }

// One classifier per input surface. It owns no session or view. Hosts must release
// remote held keys and reset routing when changing modifiers or losing focus.
@MainActor public final class NativeShortcutState {
  private let handle: NativeHandle
  public init(modifiers: NativeShortcutModifiers = .builtIn) throws {
    var raw: UInt64 = 0
    try checked { tidyvnc_shortcut_create(modifiers.rawValue, &raw, $0) }
    handle = NativeHandle(adopting: raw)
  }
  public func setModifiers(_ value: NativeShortcutModifiers) throws {
    try checked { tidyvnc_shortcut_modifiers(handle.raw, value.rawValue, $0) }
  }
  public func reset() throws { try checked { tidyvnc_shortcut_reset(handle.raw, $0) } }
  public func key(id: Int32, keysym: UInt32, down: Bool) throws -> NativeShortcutAction {
    var action: UInt32 = 0
    try checked { tidyvnc_shortcut_key(handle.raw, id, keysym, down ? 1 : 0, &action, $0) }
    guard let value = NativeShortcutAction(rawValue: action) else { throw NativeError(.internalFailure, "Invalid shortcut action") }
    return value
  }
}

// The retained viewer's Space bypass and command selection, ready for native
// desktop dispatch. A decision describes effects; it never changes wire state.
public struct NativeShortcutDecision: Equatable, Sendable {
  public enum Route: Equatable, Sendable { case remote, suppress, releaseKeyboard, captureKeyboard, contextMenu, toggleFullscreen }
  public let route: Route
  public let releaseRemoteKeys: Bool
  init(_ route: Route, release: Bool = false) { self.route = route; releaseRemoteKeys = release }
}
@MainActor public final class NativeShortcutRouter {
  private let state: NativeShortcutState
  private var pressed = Set<Int32>()
  private var bypass = false, active = false
  public init(modifiers: NativeShortcutModifiers = .builtIn) throws { state = try NativeShortcutState(modifiers: modifiers) }
  public func setModifiers(_ modifiers: NativeShortcutModifiers) throws {
    try state.setModifiers(modifiers); pressed.removeAll(); bypass = false; active = false
  }
  public func reset() throws { try state.reset(); pressed.removeAll(); bypass = false; active = false }
  // Candidates are ordered layout translations, including unmodified variants.
  // The host resolves them from the physical key, not only its modified text.
  public var routesKeyEquivalents: Bool { active || bypass }
  public func press(id: Int32, keysym: UInt32, candidates: @autoclosure () -> [UInt32]) throws -> NativeShortcutDecision {
    // Bound bypass bookkeeping too: the shared classifier is idle in that mode.
    guard pressed.contains(id) || pressed.count < 1024 else { throw NativeError(.resourceLimit, "Shortcut key capacity exceeded") }
    let action = bypass ? NativeShortcutAction.normal : try state.key(id: id, keysym: keysym, down: true)
    pressed.insert(id)
    switch action {
    case .ignore: return .init(.suppress)
    case .shortcut:
      let symbol = candidates().first { [0x20,0x47,0x67,0x4d,0x6d,0xff0d,0xff8d].contains($0) }
      if symbol == 0x20 {
        if !active { try state.reset(); bypass = true }
        return .init(.suppress)
      }
      active = true
      switch symbol {
      case 0x47,0x67: return .init(.captureKeyboard, release: true)
      case 0x4d,0x6d: return .init(.contextMenu, release: true)
      case 0xff0d,0xff8d: return .init(.toggleFullscreen, release: true)
      default: return .init(.suppress, release: true)
      }
    default: return .init(.remote)
    }
  }
  public func release(id: Int32) throws -> NativeShortcutDecision {
    let action = bypass ? NativeShortcutAction.normal : try state.key(id: id, keysym: 0, down: false)
    pressed.remove(id)
    if pressed.isEmpty { active = false }
    switch action {
    case .ignore, .shortcut: return .init(.suppress)
    case .unarm: return .init(.releaseKeyboard, release: true)
    default:
      if pressed.isEmpty { bypass = false }
      return .init(.remote)
    }
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeShortcuts.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

/// <summary>
/// Viewer shortcut modifiers. The bits are the core's; on Windows "Control"
/// is Ctrl, "Option" is Alt and "Command" is the Windows key (UX.md section 7).
/// </summary>
[Flags]
public enum NativeShortcutModifiers : uint
{
    None = 0, Control = 1, Shift = 2, Option = 4, Command = 8,
    BuiltIn = Control | Option,
}

public enum NativeShortcutAction : uint { Normal, Unarm, Shortcut, Ignore }

/// <summary>One core shortcut classifier per input surface (NativeShortcutState).</summary>
public sealed class NativeShortcutState : IDisposable
{
    private readonly NativeHandle handle;

    public unsafe NativeShortcutState(NativeShortcutModifiers modifiers = NativeShortcutModifiers.BuiltIn)
    {
        ulong raw = 0;
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_shortcut_create((uint)modifiers, &raw, &error), &error);
        handle = NativeHandle.Adopt(raw);
    }

    public unsafe void SetModifiers(NativeShortcutModifiers value)
    {
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_shortcut_modifiers(handle.Raw, (uint)value, &error), &error);
    }

    public unsafe void Reset()
    {
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_shortcut_reset(handle.Raw, &error), &error);
    }

    public unsafe NativeShortcutAction Key(int id, uint keysym, bool down)
    {
        uint action = 0;
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_shortcut_key(handle.Raw, id, keysym, down ? 1u : 0u, &action, &error), &error);
        if (!Enum.IsDefined((NativeShortcutAction)action)) throw new NativeError(NativeStatus.InternalFailure, "Invalid shortcut action");
        return (NativeShortcutAction)action;
    }

    public void Dispose() => handle.Dispose();
}

public readonly record struct NativeShortcutDecision(NativeShortcutDecision.RouteKind Route, bool ReleaseRemoteKeys = false)
{
    public enum RouteKind { Remote, Suppress, ReleaseKeyboard, CaptureKeyboard, ContextMenu, ToggleFullscreen }
}

/// <summary>The retained viewer's Space bypass and command selection (NativeShortcutRouter).</summary>
public sealed class NativeShortcutRouter : IDisposable
{
    private readonly NativeShortcutState state;
    private readonly HashSet<int> pressed = [];
    private bool bypass, active;

    public NativeShortcutRouter(NativeShortcutModifiers modifiers = NativeShortcutModifiers.BuiltIn) => state = new NativeShortcutState(modifiers);

    public void SetModifiers(NativeShortcutModifiers modifiers)
    {
        state.SetModifiers(modifiers); pressed.Clear(); bypass = false; active = false;
    }

    public void Reset()
    {
        state.Reset(); pressed.Clear(); bypass = false; active = false;
    }

    /// <summary>True while a chord or bypass owns keys (accelerators must then stay remote).</summary>
    public bool RoutesKeyEquivalents => active || bypass;

    /// <summary>Candidates are ordered layout translations of the physical key, including unmodified variants.</summary>
    public NativeShortcutDecision Press(int id, uint keysym, Func<IReadOnlyList<uint>> candidates)
    {
        if (!pressed.Contains(id) && pressed.Count >= 1024) throw new NativeError(NativeStatus.ResourceLimit, "Shortcut key capacity exceeded");
        var action = bypass ? NativeShortcutAction.Normal : state.Key(id, keysym, true);
        pressed.Add(id);
        switch (action)
        {
            case NativeShortcutAction.Ignore: return new(NativeShortcutDecision.RouteKind.Suppress);
            case NativeShortcutAction.Shortcut:
                uint[] commands = [0x20, 0x47, 0x67, 0x4d, 0x6d, 0xff0d, 0xff8d];
                var symbol = candidates().FirstOrDefault(commands.Contains);
                if (symbol == 0x20)
                {
                    if (!active) { state.Reset(); bypass = true; }
                    return new(NativeShortcutDecision.RouteKind.Suppress);
                }
                active = true;
                return symbol switch
                {
                    0x47 or 0x67 => new(NativeShortcutDecision.RouteKind.CaptureKeyboard, true),
                    0x4d or 0x6d => new(NativeShortcutDecision.RouteKind.ContextMenu, true),
                    0xff0d or 0xff8d => new(NativeShortcutDecision.RouteKind.ToggleFullscreen, true),
                    _ => new(NativeShortcutDecision.RouteKind.Suppress, true),
                };
            default: return new(NativeShortcutDecision.RouteKind.Remote);
        }
    }

    public NativeShortcutDecision Release(int id)
    {
        var action = bypass ? NativeShortcutAction.Normal : state.Key(id, 0, false);
        pressed.Remove(id);
        if (pressed.Count == 0) active = false;
        switch (action)
        {
            case NativeShortcutAction.Ignore or NativeShortcutAction.Shortcut: return new(NativeShortcutDecision.RouteKind.Suppress);
            case NativeShortcutAction.Unarm: return new(NativeShortcutDecision.RouteKind.ReleaseKeyboard, true);
            default:
                if (pressed.Count == 0) bypass = false;
                return new(NativeShortcutDecision.RouteKind.Remote);
        }
    }

    public void Dispose() => state.Dispose();
}
